cmd/transcriber/call/live_captions.go

```diff
@@ -0,0 +1,461 @@
+			}
+
+			// While audio is being transcribed, we need to cut down the window if it's > maxWindowSize.
+			window, prevTranscribedPos = cutWindowToSize(ctx.trackID, window, segments, prevTranscribedPos)
+			prevWindowLen = len(window)
+
+			// Use a for loop and a select so that we can drop ticks waiting for the transcriber.
+		waitForTranscription:
+			for {
+				select {
+				case <-ticker.C:
+					slog.Debug("processLiveCaptionsForTrack: dropped a tick waiting for the transcriber",
+						slog.String("trackID", ctx.trackID))
+				case text := <-transcribedCh:
+					if len(text) == 0 {
+						// Either transcribedCh was closed above (captionQueueCh full), or audio transcription failed.
+						// Note: this appears to happen when the transcriber fails to decode a block of audio.
+						// Usually the probability returned for the language is very low, which makes sense.
+						slog.Debug("processLiveCaptionsForTrack: received empty text, ignoring.")
+						break waitForTranscription
+					}
+					if err := t.client.SendWs(wsEvCaption, public.CaptionMsg{
+						SessionID:     ctx.sessionID,
+						UserID:        ctx.user.Id,
+						Text:          text,
+						NewAudioLenMs: float64(newAudioLenMs),
+					}, false); err != nil {
+						slog.Error("processLiveCaptionsForTrack: error sending ws captions",
+							slog.String("err", err.Error()),
+							slog.String("trackID", ctx.trackID))
+					}
+
+					break waitForTranscription
+				}
+			}
+		}
+	}
+}
+
+type segmentSamples struct {
+	Start   int
+	End     int
+	Silence bool
+}
+
+// convertToSegmentSamples turns the speech.Segments (in time) into segmentSamples (measured in samples)
+func convertToSegmentSamples(segments []speech.Segment, audioLen int) []segmentSamples {
+	var ret []segmentSamples
+	lastEndAtIdx := 0
+	for _, seg := range segments {
+		start := int(seg.SpeechStartAt * trackOutAudioRate)
+		end := int(seg.SpeechEndAt * trackOutAudioRate)
+		ret = append(ret, segmentSamples{
+			Start:   lastEndAtIdx,
+			End:     start,
+			Silence: true,
+		})
+		ret = append(ret, segmentSamples{
+			Start:   start,
+			End:     end,
+			Silence: false,
+		})
+		lastEndAtIdx = end
+	}
+
+	if lastEndAtIdx < audioLen {
+		ret = append(ret, segmentSamples{
+			Start:   lastEndAtIdx,
+			End:     audioLen,
+			Silence: true,
+		})
+	}
+
+	return ret
+}
+
+// removeShortSpeeches removes small sections of speech because either they are not actual words,
+// or the transcriber will have trouble with such a short amount.
+func removeShortSpeeches(segments []segmentSamples) {
+	for i, seg := range segments {
+		if !seg.Silence && (seg.End-seg.Start) < minSpeechLengthSamples {
+			segments[i].Silence = true
+		}
+	}
+}
+
+func cleanAudio(audio []float32, segments []segmentSamples) []float32 {
+	cleaned := append([]float32(nil), audio...)
+	for _, seg := range segments {
+		if seg.Silence {
+			for i := seg.Start; i < seg.End; i++ {
+				cleaned[i] = 0
+			}
+		}
+	}
+
+	return cleaned
+}
```

Comment:
Just pointing out that this approach makes us even more reliant on speech detector's quality and reliability.

Reply:
I suppose, but after a lot of testing it's a significant decrease in garbage output, so it definitely helps even with a spotty VAD.